src/Terminal/ui/TerminalInput.tsx

```diff
@@ -237,6 +257,28 @@ export function TerminalInput(): React.ReactElement {
       if (len == 0) {
         return;
       }
+
+      // If there is a partial command in the terminal, hitting up will filter the history
+      if(value && !autofilledValue && Settings.EnableHistorySearch) {
+        let results = searchResults;
+
+        let index;
+        if (searchTerm !== value) {
+          results = Terminal.commandHistory.filter(item => item?.startsWith(value));
```

Comment:
When testing with an artificial history of 10 million items, and with the desktop game open and many other chrome tabs, this search took ~140 ms +- 20ms.